posts/using-ansible-and-ecs-fargate-for-an-easy-client-handoff.md

```diff
@@ -0,0 +1,79 @@
+
+## Decisions that worked for us ##
+
+### Discrete front-end and back-end microservices with containers ###
+This is kind of microservices 101, but go ahead and separate front-end code early.   They will deploy and scale differently.  Segmenting also makes it easy to get truthful metrics about network traffic and system load.
+
+### Simple automated builds and redeployment with Gitlab-CI ###
+Gitlab-CI's YAML job descriptions are really quick to implement and modify.  Our job would build, test, push to repo, and call for the service to redeploy in 50 lightweight lines of YAML.  The client was able to sign up for the base-level hosted Gitlab offering, and their engineer was able to drop in the repo and CI integration.
+
+### AWS Fargate to minimize container management overhead ###
+AWS Fargate has its place.  IMHO that's for small green field projects, where there's a very basic need to run containers without persistent storage, and no other infrastructure to integrate.  Our client's project met this criteria with flying colors. Ansible 2.7 has Fargate support, which further sweetened the proposition.
+
+### Application routing delegated to ALB Rules ###
+Rather than the front end forward API requests to the backend nodes, we let ALB rules route traffic to the appropriate front or backend target group based on URL.  Service discovery was not needed since the balancer took care of routing to the different services.   Deployments instantly become transparent because ALB's healthchecks will respond to a node's availability.
+
+### Database connection information stored in AWS Systems Manger Parameter Store ###
+Since we were using Spring Boot, we had a lot of configuration methods to choose from.   Due to the small scale of the application, we opted for 12-factor style config by setting environment variables for the JDBC path and credentials.  It did not require any code changes to use.  AWS Systems Manager Parameter Store was a simple solution that could easily be extended for more configurations.  To inject the configuration, a function was added to the docker entrypoint shell script to extract the values from the parameter store and inject into the environment.
+
+
+### Config-driven ansible role to provision full AWS VPC environment and deploy ECS services ###
+Although Terraform would likely be fewer lines of code and Cloudformations would be the most AWS native, we chose to use Ansible for our VPC build.   Since the deployment code would be used by other people, we felt that Ansible code would be the easiest to read and modify.  The role was a great time server. We were able to build concurrent test environments by changing only a few configuration values.
```

Comment:
```suggestion
Although Terraform would likely be fewer lines of code and CloudFormation would be the most AWS native, we chose to use Ansible for our VPC build.   Since the deployment code would be used by other people, we felt that Ansible code would be the easiest to read and modify.  The role was a great time server. We were able to build concurrent test environments by changing only a few configuration values.
```